src/main/scala/org/apache/spark/shuffle/ucx/ShuffleTransport.scala

```diff
@@ -0,0 +1,126 @@
+/*
+* Copyright (c) 2020, NVIDIA CORPORATION. ALL RIGHTS RESERVED.
+* See file LICENSE for terms.
+*/
+package org.apache.spark.shuffle.ucx
+
+import java.nio.ByteBuffer
+
+/**
+ * Class that represents some block in memory with it's address, size.
+ *
+ * @param isHostMemory host or GPU memory
+ */
+case class MemoryBlock(address: Long, size: Long, isHostMemory: Boolean = true)
+
+/**
+ * Opaque object to describe remote memory block (address, rkey, etc.).
+ */
+trait Cookie
+
+/**
+ * Base class to indicate some blockId. It should be hashable and could be constructed on both ends.
+ * E.g. ShuffleBlockId(shuffleId: Int, mapId: Long, reduceId: Int)
+ */
+trait BlockId
+
+trait Block {
+  // Transport will call this method when it would need an actual block memory.
+  def getMemoryBlock: MemoryBlock
+}
+
+
+object OperationStatus extends Enumeration {
+  val SUCCESS, FAILURE = Value
+}
+
+/**
+ * Operation statistic, like completionTime, transport used, protocol used, etc.
+ */
+trait OperationStats {
+  def getElapsedTimeNs: Long
+}
+
+class TransportError(errorMsg: String) extends Exception(errorMsg)
+
+trait OperationResult {
+  def recvSize: Long
```

Comment:
`recvSize` doesn't make a lot of sense for `mutate` for example. 

Could we fold it into `OperationStats`? And perhaps make `OperationStats` an optional, so that it can be set to `None` if the stats don't make sense?